### PR TITLE
Profiling ctcs and fbl

### PR DESCRIPTION
**Pull request by havahol**
Okay, this became a big one! Some of it should probably be checked.

- Optimization of CTCS
- Fix of periodic boundary conditions in CTCS
- I am confident enough with the new periodic boundary conditions that I have updated the reference data for the tests accordingly. They should now be correct.
- Optimization of FBL - FBL now requires more ghost cells than before (not backward compatible)

- List of commits for FBL and CTCS in the git-version-benchmark notebook.
- Updated the run_benchmark script to account for the new data size for FBL
- The git-version-benchmark notebook now checks out the new run_benchmark script for the benchmarking. The old and new version of the script is compared, and this should only affect the new data size for FBL input (as intended)

All tests pass both on laptop and desktop

### CODE REVIEW

File: gpu_ocean/SWESimulators/CTCS.py [hunk 1]
```diff
@@ -155,6 +162,20 @@ def __init__(self, \
                                                  halo_x, halo_y \
         )
         
+         #"Beautify" code a bit by packing four bools into a single int
+        #Note: Must match code in kernel!
+        self.wall_bc = np.int32(0)
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/CTCS.py]: Why do we have this, and the next one which is very similar? (see next comment)

File: gpu_ocean/SWESimulators/CTCS.py [hunk 2]
```diff
@@ -271,47 +292,40 @@ def step(self, t_end=0.0):
...
-                    self.r, self.A,\
-                    self.H.data.gpudata, self.H.pitch, \
-                    self.gpu_data.h1.data.gpudata, self.gpu_data.h1.pitch,      # eta^{n} \
-                    self.gpu_data.hu0.data.gpudata, self.gpu_data.hu0.pitch,    # U^{n-1} => U^{n+1} \
-                    self.gpu_data.hu1.data.gpudata, self.gpu_data.hu1.pitch,    # U^{n} \
-                    self.gpu_data.hv1.data.gpudata, self.gpu_data.hv1.pitch,    # V^{n} \
-                    wind_stress_t)
+            #"Beautify" code a bit by packing four bools into a single int
+            #Note: Must match code in kernel!
+            boundary_conditions = np.int32(0)
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/CTCS.py]: What is the difference between this and the previous?

File: gpu_ocean/SWESimulators/Common.py [hunk 3]
```diff
@@ -479,34 +479,39 @@ class SWEDataArakawaC:
     We use h as cell centers
     """
     def __init__(self, gpu_stream, nx, ny, halo_x, halo_y, h0, hu0, hv0, \
-                 asymHalo=None):
+                 fbl=False):
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/Common.py]: I think we should refactor this in a subsequent PR so that SWEDataArakawaCFBL becomes a class to handle the special case. Right now it is OK, but it should be on our todo-list somewhere

File: gpu_ocean/SWESimulators/Common.py [hunk 4]
```diff
@@ -479,34 +479,39 @@ class SWEDataArakawaC:
...
-            #print(asymHalo)
-            assert(max(asymHalo) <= 1)
-            asymHaloU = [asymHalo[0], 0, asymHalo[2], 0]
-            asymHaloV = [0, asymHalo[1], 0, asymHalo[3]]
+        
+        self.fbl = fbl
+        
+        if (fbl):
+            self.h0   = CUDAArray2D(gpu_stream, nx  , ny  , halo_x, halo_y, h0)
+            self.hu0  = CUDAArray2D(gpu_stream, nx-1, ny  , halo_x, halo_y, hu0)
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/Common.py]: Is -1 here really correct?

File: gpu_ocean/SWESimulators/FBL.py [hunk 5]
```diff
@@ -311,13 +306,12 @@ class FBL_periodic_boundary:
...
                  nx, ny, \
-                 boundary_conditions, asym_ghost_cells, \
+                 boundary_conditions, \
                  block_width=16, block_height=16 ):
 
         self.boundary_conditions = boundary_conditions
-        self.asym_ghost_cells = asym_ghost_cells
-        self.ghostsX = np.int32(self.asym_ghost_cells[1] + self.asym_ghost_cells[3])
-        self.ghostsY = np.int32(self.asym_ghost_cells[0] + self.asym_ghost_cells[2])
+        self.ghostsX = np.int32(0) #self.asym_ghost_cells[1] + self.asym_ghost_cells[3])
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/FBL.py]: Is this intentional?

File: gpu_ocean/SWESimulators/FBL.py [hunk 6]
```diff
@@ -367,61 +369,39 @@ def __init__(self, \
...
-            
-            self.updateGhostCellsUKernel.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
+        if (self.boundary_conditions.east == 2):
+            self.periodicBoundaryUKernel_EW.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
                         self.nx, self.ny, \
-                        self.nx_halo, self.ny_halo, \
                         hu0.data.gpudata, hu0.pitch)
-            
+        
+        self.firstU = False
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/FBL.py]: Why are these firstU and firstV needed?

File: gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu [hunk 7]
```diff
@@ -24,60 +24,48 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
   * Kernel that evolves U one step in time.
   */
 extern "C" {
 __global__ void computeUKernel(
         //Discretization parameters
-        int nx_, int ny_,
-        int bc_east_, int bc_west_,
-        float dx_, float dy_, float dt_,
+        const int nx_, const int ny_,
+		const int wall_bc_,
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu]: Tabs used here

File: gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu [hunk 8]
```diff
@@ -206,19 +195,18 @@ __global__ void computeUKernel(
...
     //Reconstruct H_bar and H_x (at the U position)
     const float H_bar_0m = 0.25f*(H_0m + H_pm + H_00 + H_p0);
     const float H_bar_00 = 0.25f*(H_00 + H_p0 + H_0p + H_pp);
     const float H_x = 0.5f*(H_00 + H_p0);
     
-    //Reconstruct Eta_bar at the V position
-    const float eta_bar_0m = 0.25f*(eta_0m + eta_pm + eta_00 + eta_p0);
-    const float eta_bar_00 = 0.25f*(eta_00 + eta_p0 + eta_0p + eta_pp);
-
+	//Reconstruct Eta_bar at the V position
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu]: Tabs vs spaces

File: gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu [hunk 9]
```diff
@@ -228,38 +216,39 @@ __global__ void computeUKernel(
...
+	const float P_x_hat = -0.5f*g_*(eta_p0*eta_p0 - eta_00*eta_00);
+	const float P_x = -g_*h_x*(eta_p0 - eta_00) + P_x_hat;
+	
     //Calculate nonlinear effects
-    const float N_a = (U_p0 + U_00)*(U_p0 + U_00) / (H_p0 + eta_p0);
-    const float N_b = (U_00 + U_m0)*(U_00 + U_m0) / (H_00 + eta_00);
-    const float N_c = (U_0p + U_00)*(V_p0 + V_00) / (H_bar_00 + eta_bar_00);
-    const float N_d = (U_00 + U_0m)*(V_pm + V_0m) / (H_bar_0m + eta_bar_0m);
-    float N = 0.25f*( N_a - N_b + (dx_/dy_)*(N_c - N_d) );
+	const float N_a = (U_p0 + U_00)*(U_p0 + U_00) / (H_p0 + eta_p0);
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/CTCS_U_kernel.cu]: Please reformat file with spaces only (applies to all of them, I guess)

File: gpu_ocean/SWESimulators/gpu_kernels/CTCS_eta_kernel.cu [hunk 10]
```diff
@@ -98,17 +87,17 @@ __global__ void computeEtaKernel(
     }
     
     //Read V into shared memory
-    for (int j=ty; j<block_height+1; j+=blockDim.y) {
+    for (int j=ty; j<block_height+2; j+=blockDim.y) {
         //const int l = clamp(by + j - 1, 0, ny_); // prevent out of bounds
-        const int l = by + j;
+        const int l = by + j - 1;
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/CTCS_eta_kernel.cu]: I trust that these are correct, but now the kernels are really challenging wrt indices

File: gpu_ocean/SWESimulators/gpu_kernels/FBL_periodic_boundary.cu [hunk 11]
```diff
@@ -38,133 +44,217 @@ __global__ void closedBoundaryUKernel(
     //Compute pointer to current row in the U array
     float* const U_row = (float*) ((char*) U_ptr_ + U_pitch_*tj);	
     
-    if ( (ti ==0 || ti == nx_) && tj < ny_halo_) {
+    if ( ( ((ti == 0  ) && (bc_west_ == 1)) || 
+           ((ti == nx_) && (bc_east_ == 1))    ) && tj < ny_ + 2) {
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/FBL_periodic_boundary.cu]: Why  && tj < ny_ + 2  here)

File: gpu_ocean/SWESimulators/gpu_kernels/FBL_step_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,255 @@
+/*
+This OpenCL kernel implements part of the Forward Backward Linear 
+numerical scheme for the shallow water equations, described in 
+L. P. Røed, "Documentation of simple ocean models for use in ensemble
+predictions", Met no report 2012/3 and 2012/5 .
+
+Copyright (C) 2016  SINTEF ICT
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/FBL_step_kernel.cu]: The copyright dates should be updated to include 2018 also

File: gpu_ocean/demos/testCasesDemos/run_benchmark.py [hunk 13]
```diff
@@ -38,6 +38,8 @@
 parser.add_argument('--iterations', type=int, default=1)
 parser.add_argument('--simulator', type=str)
 parser.add_argument('--output', type=str, default=None)
+parser.add_argument("--new_fbl", type=int, default=0)
```
Comment (babrodtk) [inline gpu_ocean/demos/testCasesDemos/run_benchmark.py]: I think old-fbl should be removed, thus not requiring this

Comment (havahol) [reply gpu_ocean/demos/testCasesDemos/run_benchmark.py]: old-fbl is required in order to benchmark the original "untuned FBL"

File: gpu_ocean/SWESimulators/Common.py (hunk 3 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/Common.py]: Agree. Made card in trello

File: gpu_ocean/SWESimulators/Common.py (hunk 4 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/Common.py]: Yes. And I thought this looked better that `halo_x - 0.5`

File: gpu_ocean/SWESimulators/FBL.py (hunk 5 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/FBL.py]: I saw it as well, and no :) All open boundary conditions must be checked later, as they are not tested during this work. Adding this task to trello

File: gpu_ocean/SWESimulators/FBL.py (hunk 6 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/FBL.py]: Earlier:  used for debugging in open boundary conditions. 
Now: closed boundary conditions (setting velocities defined orthogonal to the boundary to zero) is handled within the step kernel. We still call the boundary kernels the prior to the first time step to be sure that this is correct (to avoid user error), but do not call the closed boundary kernels afterwards.

File: gpu_ocean/demos/testCasesDemos/run_benchmark.py (hunk 13 above)
Comment (babrodtk) [reply gpu_ocean/demos/testCasesDemos/run_benchmark.py]: Isn't that a git checkout away? That's what we do with the other ones, right?

File: gpu_ocean/SWESimulators/gpu_kernels/CTCS_eta_kernel.cu (hunk 10 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/gpu_kernels/CTCS_eta_kernel.cu]: Yes, and indices are updated correspondingly

File: gpu_ocean/SWESimulators/gpu_kernels/FBL_periodic_boundary.cu (hunk 11 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/gpu_kernels/FBL_periodic_boundary.cu]: So that we don't write outside the buffer... why else? :)

**Comment by havahol** [comment]
All comments followed or replied to
